sklearn/base.py

```diff
@@ -271,9 +271,20 @@ def __repr__(self, N_CHAR_MAX=700):
         return repr_
 
     def __getstate__(self):
+        if hasattr(self, "__slots__"):
```

Comment:
There's a difference between `getattr` and `hasattr` here. We should do `getattr` since if the attribute exists but it's none, we shouldn't fail.

Reply:
> if the attribute exists but it's none, we shouldn't fail.

Isn't that what's happening with the existing code?

```python
In [1]: class A:
   ...:     def __init__(self):
   ...:         self.x = None
   ...: 

In [2]: hasattr(A(), "x")
Out[2]: True
```

Reply:
Also, it seems that you can't do `__slots__ = None` in the first place.

Reply:
@adrinjalali I think that `__slots__` cannot be `None`:

```python
Cell In[33], line 1
----> 1 class A:
      2     __slots__ = None

TypeError: 'NoneType' object is not iterable
```

Reply:
Sure, but if the code in cpython does `getattr`, I rather stay on that safer side. I don't know why they did that, and I'm not sure if I wanna know why lol

Reply:
The only difference I can think of:

```python
>>> class A:
...   __slots__ = []
... 
>>> a = A()
>>> hasattr(a, "__slots__")
True
>>> bool(getattr(a, "__slots__", None))
False
```

Honestly, `hasattr` looks more reasonable to me, not sure what the intent is of using `getattr` in `copyreg`.

Reply:
I think there can be a C extension type which exposes a `__slots__=None`, and that's why `getattr(obj, '__slots__', None)` is more correct than `getattr(obj, '__slots__')`.

Reply:
But what about `__slots__ = []`, that should also trigger the error. We would have to additionally check for empty list and empty tuple.

Reply:
I wouldn't try to be more correct than cpython, and I'm not sure what `__slot__=[]` would mean. I can understand `__slots__=None` would be equivalent to it not existing.

Reply:
On having to understand the subtle differences between `getattr` and `hasattr`: I'm with Adrin on the "there must be a reason the CPython devs used that, even if we don't fully understand why".

Reply:
Okay, I changed to `if getattr(self, "__slots__", None)`